src/index/CompressedBlockPrefiltering.h

```diff
@@ -0,0 +1,131 @@
+//  Copyright 2024, University of Freiburg,
+//                  Chair of Algorithms and Data Structures
+//  Author: Hannes Baumann <baumannh@informatik.uni-freiburg.de>
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "global/Id.h"
+#include "global/ValueIdComparators.h"
+#include "index/CompressedRelation.h"
+
+namespace prefilterExpressions {
+
+// The compressed block metadata (see `CompressedRelation.h`) that we use to
+// filter out the non-relevant blocks by checking their content of
+// `firstTriple_` and `lastTriple_` (`PermutedTriple`)
+using BlockMetadata = CompressedBlockMetadata;
+
+//______________________________________________________________________________
+/*
+`PrefilterExpression` represents a base class for the following sub-classes that
+implement the block-filtering procedure for the specific relational + logical
+operations/expressions.
+
+Remark: We do not actually evaluate the respective SPARQL Expression. We only
+pre-filter w.r.t. blocks that contain relevant data for the actual evaluation of
+those expressions to make the evaluation procedure more efficient.
+
+The block-filtering is applied with the following operations:
+Relational Expressions - `<=`, `>=`, `<`, `>`, `==` and `!=`
+Logical Operations - `and` and `or`
+*/
+
+class PrefilterExpression {
+ public:
+  virtual ~PrefilterExpression() = default;
+
+  virtual std::unique_ptr<PrefilterExpression> logicalComplement() const = 0;
+
+  // The respective metadata to the blocks is expected to be provided in
+  // a sorted order (w.r.t. the relevant column).
+  virtual std::vector<BlockMetadata> evaluate(
+      const std::vector<BlockMetadata>& input,
+      size_t evaluationColumn) const = 0;
+};
+
+//______________________________________________________________________________
+// For the actual comparison of the relevant ValueIds from the metadata triples,
+// we use the implementations from ValueIdComparators.
+//
+// Supported comparisons are:
+//  - LessThan, LessEqual, Equal, NotEqual, GreaterEqual, GreaterThan
+using CompOp = valueIdComparators::Comparison;
+
+//______________________________________________________________________________
+template <CompOp Comparison>
+class RelationalExpression : public PrefilterExpression {
+ private:
+  // The ValueId on which we perform the relational comparison on.
+  ValueId referenceId_;
+
+ public:
+  explicit RelationalExpression(const ValueId referenceId)
+      : referenceId_(referenceId) {}
+
+  std::unique_ptr<PrefilterExpression> logicalComplement() const override;
+
+  std::vector<BlockMetadata> evaluate(const std::vector<BlockMetadata>& input,
+                                      size_t evaluationColumn) const override;
+};
+
+//______________________________________________________________________________
+// Helper struct for a compact class implementation regarding the logical
+// operations `AND`, `OR` and `NOT`.
+enum struct LogicalOperators { AND, OR, NOT };
```

Comment:
I think NOT is neither needed nor implemented.

Reply:
let me see.

Reply:
I believe `NOT` is simply a separate class, because
1. It has only one child
2. It does something completely different.